keps/sig-windows/2258-node-service-log-viewer/README.md

```diff
@@ -0,0 +1,373 @@
+that is lacking a client. Given its existence we can supplement that with a
+wafer thin shim over the /journal directory that shells out to journalctl. This
+allows us to extend the endpoint for getting logs from the system journal on
+Linux systems that support systemd. To enable filtering of logs, we can reuse
+the existing filters supported by journalctl. The `kubectl logs` will have
+command line options for specifying these filters when interacting with node
+objects.
+
+On the Windows side viewing of logs from services that use `C:\var\log` will
+be supported by the existing endpoint. For Windows services that log to the
+the System and Application logs, Windows Event Logs and Event Tracing (ETW),
+we can leverage the [Get-WinEvent cmdlet](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.diagnostics/get-winevent?view=powershell-7.1)
+that supports getting logs from all these sources. The cmdlet has filtering
+options that can be leveraged to filter the logs in the same manner we do
+with the journal logs.
+
+Please note that filtering will not be available for logs in `/var/log/` or
+`C:\var\log\`.
+
+On clusters where the same service using different logging mechanisms,
+example: kubelet logs to `/var/log` on some nodes and uses the journal on other
+nodes, the cluster administrator will have to use `kubectl logs` with different
+set of options to read both types of logs.
+
+The feature now enables the cluster administrator to interrogate all services.
+This could be prevented by having a whitelist of allowed services. But this
+comes with severe disadvantages as there could be nodes (especially with
+Windows) that have other services to support networking and monitoring.
+These services are variable and will depend on how the nodes have been
+configured. Here are some examples:
+- (hybrid-overlay-node)[https://github.com/ovn-org/ovn-kubernetes/tree/master/go-controller/hybrid-overlay]
+- (windows-exporter)[https://github.com/prometheus-community/windows_exporter].
+
+When introducing this feature it will be hidden behind a `NodeLogs` feature
+gate in the kubelet that needs to be explicitly enabled. To view logs not
+present in `/var/log/` or `C:\var\log\` will require this feature gate to
+be enabled. This is because the `/var/log/` endpoint has been present for over
+5 years and it is possible that a client is available in the wild. Given that
+we cannot hide that feature behind a feature gate.
+
+A reference implementation of this feature without the feature gate is
+available [here](https://github.com/kubernetes/kubernetes/pull/96120).
+
+#### kubectl
+
+`kubectl` has an existing `logs` command that is used to view the logs for a
+container in a pod or a specified resource. The sub-command looks at resource
+types, so can be extended to work with node objects to view the logs of services
+on the nodes. Given that the `logs` command depends on RBAC policies for access
```

Comment:
We need to add a section on RBAC policies. Who has access?

Reply:
As per @soltysh the existing logs command depend on RBAC policies for access to each resource it supports and it would just extend to nodes. @soltysh please correct me if I misunderstood or if I need to clarify that more in this section.

Reply:
That's how I understand the code as well. The `logs` cmd solely constructs and sends an API call (currently referencing `pods` by default) - it doesn't concern itself with RBAC at all. If the objects referenced in the API call can't be accessed by the user, because there is an RBAC policy saying so, the API server will return an error.

Reply:
Subresources have to be granted to users (you do not get subresourec access by default), so a user who can GET or LIST nodes does not automatically have access to GET nodes/logs